content/Building an Engine-modularized API in Rails - Guillermo Leopold.md

```diff
@@ -0,0 +1,154 @@
+  * Copy a generic version spec_helper, rails_helper.
+  * Rename/adapt some filenames to the standard that we are following.
+
+I won’t go deeper on which files we removed or we added as this may change according to the needs of each project but I would like to mention a few configurations that are needed to make this work properly and are going to save you a couple of headaches.
+
+  1. Related to the migrations, we want each engine to have it owns migrations files, to achieve this we added the following block of code in lib/engine_name/engine.rb:
+
+```Ruby
+initializer :append_migrations do |app|
+  unless app.root.to_s.match? root.to_s
+    config.paths['db/migrate'].expanded.each do |expanded_path|
+      app.config.paths['db/migrate'] << expanded_path
+    end
+  end
+end
+```
+
+So, with this code we are telling Rails to look for new migrations in the db/migrate folder that belongs to the engine.
+
+  2. We want to be able to define a routes file in each engine and have the ability to have access to those endpoints from the main app. An easy way to achieve this and keeping it generic, is adding this piece of code in the routes.rb in the main application
+
+```Ruby
+Dir.glob(File.expand_path('../engines/*', __dir__)).each do |path|
+  engine = File.basename(path)
+  mount engine.classify.constantize::Engine, at: '/'
+end
+```
+
+This is iterating over the engines directory and mounting the routes of each engine.
+
+  3. A similar approach is taken to load all the engines in the gemfile in the main application.
+
+```Ruby
+Dir.glob(File.expand_path('../engines/*', __FILE__)).each do |path|
+  engine = File.basename(path)
+  gem engine, path: "engines/#{engine}", require: (ENV['ENGINE'].nil? || ENV['ENGINE'] == engine)
+end
+```
+
+By creating the script and adding these configurations into the main application, we were able to simplify a lot of the processes to add new engines and we ensured that all the engines are going to follow the same standard.
+
+
+## Developing and testing a feature: considerations and tips
+
+Now that we have gone through all this process that may seem extra thorough at the beginning, we are ready to start.
+At this point, if everything is ok, there shouldn’t be many problems or inconveniences as we are going to be doing pretty much what we already used to do while developing features in Rails.
+
+There are still a few considerations related to how engines work that are worth mentioning:
+  * An engine works like a library, so when adding dependencies in between, we must take care to avoid circular dependencies. This means that if we have the engines A, B and C; if the engine A can see B and B can see C, then A may be able to see C but C shouldn’t be able to see either B or A.
+  * Following on from this, when adding a dependency between engines, we need to add this one into the gemfile that belongs to that engine, not the gemspec. We will also need to add the engines that the other engine has as dependencies too.
```

Comment:
Mabye you could add some info/text about what happens when a engines that cannot see another needs to interact with it. You can talk about Observer, show some example, and maybe also recommend using async communication patterns like Queue pub/sub or similar

Reply:
Good one